edb/pgsql/compiler/relgen.py

```diff
@@ -2003,6 +2008,143 @@ def process_set_as_existence_assertion(
     return new_stmt_set_rvar(ir_set, stmt, ctx=ctx)
 
 
+def process_set_as_multiplicity_assertion(
+    ir_set: irast.Set,
+    stmt: pgast.SelectStmt,
+    *,
+    ctx: context.CompilerContextLevel,
+) -> SetRVars:
+    """Implementation of std::assert_distinct"""
+    expr = ir_set.expr
+    assert isinstance(expr, irast.FunctionCall)
+
+    ir_arg = expr.args[0]
+    ir_arg_set = ir_arg.expr
+
+    # Generate a distinct set assertion as the following SQL:
+    #
+    #    SELECT
+    #        <target_set>,
+    #        (CASE WHEN
+    #            <target_set>
+    #            IS DISTINCT FROM
+    #            lag(<target_set>) OVER (ORDER BY <target_set>)
+    #        THEN <target_set>
+    #        ELSE edgedb.raise(ConstraintViolationError)) AS check_expr
```

Comment:
Should we add a new class of errors -- RuntimeAssertionError?

Reply:
Why would that be better than a `ConstraintViolationError` or a `CardinalityViolationError`?